File: bfv/ntt.py
from poly import Poly
from bb import bmm

def main():
  size = 256
  q = 3329
  psi = 17


  '''
  size = 2**11
  q = 274_877_908_993
  psi = 124878001
  '''

  n = 2
  size *= 2
  n *= 2

  invpsi = pow(psi, -1, q)

  fn = Poly([1] + ([0]*(size-1)) + [1])  

  a = Poly([1, 2, 3, 4])
  a = Poly(list(range(size)))
  b = Poly([0, 2, 3, 4])

  z = a*b
  quo, rem = z // fn

  psi_powers = []
  invpsi_powers = []
  for i in range(size//n):
    psi_powers.append( pow(psi, bitrev(i, size//n), q) )
    invpsi_powers.append( pow(invpsi, bitrev(i, size//n), q) )
  
  psi_poly = []
  powers = []
  for i in range(size//n):
    powers.append( pow(psi, 2*bitrev(i, size//n) + 1, q) )
    p = Poly([-1*(pow(psi, 2*bitrev(i, size//n) + 1, q)), 0, 1])
    p = p % q
    psi_poly.append(p)
  prod = Poly([1])
  for ind, p in enumerate(psi_poly):
    prod = prod * p
    prod = prod % q

  na = NTT(a, size, q, psi_powers, n)
  na = na % q
  ia = iNTT(na, size, q, psi_powers, n)
  ia = ia % q
  # print(ia)
  
  nb = NTT(b, size, q, psi_powers, n)
  nb = nb % q
  ib = iNTT(nb, size, q, psi_powers, n)
  # print(ib)

  # c = [ (xa * xb) % q for xa, xb in zip(na, nb) ]
  # print(powers)
  c = ntt_mult(na, nb, size, q, powers, n)
  ic = iNTT( c, size, q, psi_powers, n )
  rem = rem % q
  print(rem)
  print()
  print(ic)
  print()
  print( ic == rem )
  return

  print( powers[:64] )
  test = Poly( list(range(size)) )
  zest = NTT(test, size, q, psi_powers)
  zest = zest % q
  mest = ntt_mult(zest, zest, size, q, powers)
  # print(mest.poly[:64])
  best = iNTT( mest, size, q, psi_powers)
  z = test * test
  quo, rem = z // fn
  rem = rem % q
  print(best.poly[:10] )
  print( rem == best )

  return

def ntt_mult(poly_a, poly_b, size, q, psi, nsz=2):
  a = poly_a.poly.copy()
  b = poly_b.poly.copy()

  c = []

  for i in range(0, size//nsz):
    di = i*nsz
    # psi_p = pow(psi, 2*bitrev(i, size//2)+1, q)
    psi_p = psi[i]
    c += __testmul(a[di:di+nsz], b[di:di+nsz], q, psi_p, nsz)
  return Poly(c)

def basemul(poly_a, poly_b, q, psi):
  r = [0, 0]
  r[0]  = (poly_a[1] * poly_b[1]) % q
  r[0]  = (r[0] * psi) % q
  r[0] += (poly_a[0] * poly_b[0])
  r[0] %= q
  r[1]  = (poly_a[0] * poly_b[1]) % q
  r[1] += (poly_a[1] * poly_b[0]) % q
  r[1] %= q
  return r

def testmul(poly_a, poly_b, q, psi, size=2):
  r = [0 for _ in range(size)]
  for i in range(size):
    for j in range(size):
      r[(i+j)%size] += poly_a[i] * poly_b[j] if i+j < size else ( poly_a[i] * poly_b[j] * psi )
  r = [ri % q for ri in r]
  return r

def __testmul(poly_a, poly_b, q, psi, size=2):
  r = [0 for _ in range(size)]
  for i in range(1, size):
    for j in range(size - i, size):
      r[i+j-size] += (poly_a[i]*poly_b[j]*psi)
  for i in range(size):
    for j in range(size-i):
      r[i+j] += (poly_a[i] * poly_b[j])
  r = [ri % q for ri in r]
  return r

def NTT(poly, size, mod, psi_powers, n=2):
  r = poly.copy()
  k = 1
  length = size // 2
  # length starts at half, and goes to n where (x^n - psi)
  while length >= n:
    start = 0
    while start < size:
      zeta = psi_powers[k]
      k += 1
      for j in range(start, start+length):
        # t = (zeta * r[j + length]) % mod
        t = bmm(zeta, r[j + length], mod)
        r[j + length] = r[j] - t
        r[j] = r[j] + t
        r[j + length] %= mod
        r[j] %= mod
      start = start + length + length
    length = length >> 1
  return r
  '''
  ret = [0]*size
  f = poly.copy()
  half = size//2
  for i in range(half):
    s0 = 0
    s1 = 0
    for j in range(half):
      index = (2*bitrev(i, half) + 1)*j
      index %= size
      s0 += (f[2*j])*psi_powers[index]
      s1 += (f[2*j+1])*psi_powers[index]
    ret[2*i] = s0
    ret[2*i + 1] = s1
  return Poly(ret)   
  '''

def iNTT(poly, size, mod, psi_powers, n=2):
  r = poly.copy()

  k = (size//n) - 1
  length = n
  # length starts at n where (x^n - psi)
  # length ends at half
  while length <= (size//2):
    start = 0
    while start < size:
      zeta = psi_powers[k]
      k -= 1
      for j in range(start, start + length):
        t = r[j]
        r[j] = (t + r[j + length]) % mod
        r[j + length] = r[j + length] - t
        r[j + length] = (zeta*r[j + length]) % mod
      start = start + length + length
    length = length << 1

  invsize = pow(size//n, -1, mod)
  # invsize = 1175
  # invsize = 1441
  for i in range(size):
    r[i] = (r[i]*invsize) % mod
    pass
  return r   

def bitrev(k: int, size: int):
  # return bit-reversed order of k
  
  revk = 0
  bit_len = (size-1).bit_length()

  for i in range( bit_len ):
    revk += (2 ** (bit_len-1-i) ) * ( ((2**i) & k ) >> i )

  return revk

def __NTT(p: Poly, size, mod, psi):
  # this will convert p into NTT

  a = p.copy()
  # ret = Poly( ret.poly + ([0]*(self.n-len(p))) )

  m = 1
  k = size // 2
  while m < size:
    for i in range(m):
      jFirst = 2 * i * k
      jLast = jFirst + k
      wi = psi[ m+i ]
      for j in range(jFirst,jLast):
        l = j + k
        t = a[j]
        u = a[l] * wi
        a[j] = (t + u) % mod
        a[l] = (t - u) % mod

    m = m * 2
    k = k//2

  return a 


if __name__ == '__main__':
  main()

File: bfv/ntt.c
#include "ntt.h"

i64* NTT(i64* poly, u64 size, i64 mod, i64* psi_powers) {
  i64 i, j;

  i64* res = copy_poly(poly, size);
  poly_mod(res, size, mod);
  // i64* res = poly;

  i64 m, k, jstart, jend, wi, t, u, l;

  m = 1;
  k = size>>1;
  for (;m < size; m = m<<1) {
    for (i = 0; i < m; i++) {
      jstart = 2 * i * k;
      jend = jstart + k;
      wi = psi_powers[ m+i ];
      for (j = jstart; j < jend; j++) {
        l = j + k;
        t = res[j] % mod;
        u = big_mult_mod((u64)res[l], (u64)wi, (u64)mod);
        /*
        if ((m == 256) && (j % 2 == 0)) {
          printf("m: %ld\ti: %ld\tj: %ld\tu: %ld\twi: %ld\tres[l]: %ld\n", m, i, j, u, wi, res[l]);
          // printf("k: %ld\n", k);
        }
        */
        res[j] = (t + u) % mod;
        res[l] = (t - u) % mod;
        if (res[j] < 0)   res[j] += mod;
        if (res[j+k] < 0) res[j+k] += mod;
      }
    }
    k = k>>1;
  }

  poly_mod(res, size, mod);

  return res;
}

i64* iNTT(i64* poly, u64 size, i64 mod, i64* invpsi_powers) {
  i64 i, j;

  i64* res = copy_poly(poly, size);

  i64 m, k, jstart, jend, wi, t, u, l, temp;

  m = size>>1;
  k = 1;
  for (;m >= 1; m = m>>1) {
    for (i = 0; i < m; i++) {
      jstart = 2 * i * k;
      jend = jstart + k;
      wi = invpsi_powers[ m+i ];
      for (j = jstart; j < jend; j++) {
        l = j + k;
        t = res[j] % mod;
        u = res[l] % mod;
        res[j]   = (t + u) % mod;
        temp = (t-u);
        if (temp < 0) temp += mod;
        res[l] = big_mult_mod((u64)temp, (u64)wi, (u64)mod);
        if (res[j] < 0) res[j] += mod;
        if (res[l] < 0) res[l] += mod;
      }
    }
    k = k<<1;
  }

  i64 invn = inverse_mod(size, mod);
  for (i = 0; i < size; i++) {
    res[i] = big_mult_mod((u64)res[i], (u64)invn, (u64)mod);
  }
  poly_mod(res, size, mod);

  return res;
}

i64* prepare_psi_powers(u64 size, i64 mod, i64 psi) {
  i64 i, j;
  i64* psi_powers = (i64*) malloc( sizeof(i64) * size );

  i64 logn = 0;
  i = 1;
  for (; i < size; i = i<<1)
    logn++;
  
  for (i = 0; i < size; i++) {
    psi_powers[i] = 1;
    for (j = 0; j < bitrev(i, logn); j++) {
      psi_powers[i] = big_mult_mod((u64)psi_powers[i], (u64)psi, (u64)mod);
    }
  }
  return psi_powers;
}

i64* prepare_invpsi_powers(u64 size, i64 mod, i64 invpsi) {
  i64 i, j;
  i64* invpsi_powers = (i64*) malloc( sizeof(i64) * size );

  i64 logn = 0;
  i = 1;
  for (; i < size; i = i<<1)
    logn++;

  for (i = 0; i < size; i++) {
    invpsi_powers[i] = 1;
    for (j = 0; j < bitrev(i, logn); j++) {
      invpsi_powers[i] = big_mult_mod((u64)invpsi_powers[i], (u64)invpsi, (u64)mod);
    }
  }
  return invpsi_powers;
}

i64 inverse_mod(i64 a, i64 q) {
  i64 t, nt, r, nr, temp, quo;

  t  = 0;
  nt = 1;
  r  = q;
  nr = a;

  while (nr != 0) {
    quo = r / nr;
    temp = nt;
    nt = (t - (quo * temp));
    t = temp;

    temp = nr;
    nr = (r - (quo * temp));
    r = temp;
  }

  if (t < 0)
    t += q;

  return t;
}

i64 bitrev(i64 num, u64 bitsize) {
  int i;
  i64 reversed = 0;
  for (i = 0; i < bitsize; i++) {
    if (num % 2 == 1) reversed++;
    reversed = reversed << 1;
    num = num >> 1;
  }

  return (reversed >> 1);
}

i64 big_mult_mod(u64 x, u64 y, u64 mod) {
  if ( (x < 0x100000000) && (y < 0x100000000) ) {
    return (x*y)%mod;
  }

  u64 x_low, x_high, y_low, y_high;
  u64 mask;
  mask = 0xffffffff;

  x_low  = (x&mask);
  x_high = (x>>16);
  x_high = (x_high>>16);

  y_low  = (y&mask);
  y_high = (y>>16);
  y_high = (y_high>>16);

  u64 z_low, z_mid, z_high;

  z_low  = x_low * y_low;
  z_mid  = (x_low * y_high) + (x_high * y_low);
  z_high = x_high * y_high;

  i64 shifting = 16;

  // shift and mod z_mid and z_high
  u64 i;
  for (i = 0; i < 32/shifting; i++) {
    z_mid  = (z_mid<<shifting) % mod;
    z_high = (z_high<<shifting) % mod;
    z_high = (z_high<<shifting) % mod;
  }

  i64 ret = (z_low%mod) + (z_mid%mod) + (z_high%mod);

  return (ret%mod);
}


void basemul(i64* r, i64* poly_a, i64* poly_b, i64 mod, i64 zeta) {
  r[0]  = big_mult_mod(poly_a[1], poly_b[1], mod);
  r[0]  = big_mult_mod(r[0], zeta, mod);
  r[0] += big_mult_mod(poly_a[0], poly_b[0], mod);
  r[0]  = (r[0] % mod);
  r[1]  = big_mult_mod(poly_a[0], poly_b[1], mod);
  r[1] += big_mult_mod(poly_a[1], poly_b[0], mod);
  r[1]  = (r[1] % mod);
  return;
}


i64* zeta_powers(u64 size, i64 mod, i64 psi, u64 nsize) {
  i64* zetas = (i64*) malloc(sizeof(i64) * size / nsize);
  i64 i, j, prod;

  i64 logn = 0;
  i = 1;
  for (; i < size/nsize; i = i<<1)
    logn++;

  for (i = 0; i < size/nsize; i++) {
    prod = 1;
    for (j = 0; j < bitrev(i, logn); j++) {
      prod = big_mult_mod(prod, psi, mod);
    }
    zetas[i] = prod;
  }
  return zetas;
}

i64* zeta_mults(u64 size, i64 mod, i64 psi, u64 nsize) {
  i64* zetas = (i64*) malloc(sizeof(i64) * size / nsize);
  i64 i, j, prod;

  i64 logn = 0;
  i = 1;
  for (; i < size/nsize; i = i<<1)
    logn++;

  for (i = 0; i < size/nsize; i++) {
    prod = 1;
    for (j = 0; j < 2*bitrev(i, logn)+1; j++) {
      prod = big_mult_mod(prod, psi, mod);
    }
    zetas[i] = prod;
  }
  return zetas;
}

i64* zNTT(i64* poly, u64 size, i64 mod, i64* zetas, u64 nsize) {
  i64 len, start, j, k, t, zeta;
  k = 1;
  i64 *r = (i64*) malloc(sizeof(i64) * size);
  memcpy(r, poly, sizeof(i64)*size );
  poly_mod(r, size, mod);
  for (len = size/2; len >= nsize; len = len >> 1) {
    for (start = 0; start < size; start = j + len) {
      zeta = zetas[k++];
      for (j = start; j < start + len; j++) {
        t = big_mult_mod(zeta, r[j + len], mod);
        r[j + len] = (r[j] - t) % mod;
        if (r[j + len] < 0) r[j + len] += mod;
        r[j] = (r[j] + t) % mod;
        if (r[j] < 0) r[j] += mod;
      }
    }
  }
  poly_mod(r, size, mod);
  return r;
}

i64* ziNTT(i64* poly, u64 size, i64 mod, i64* zetas, u64 nsize) {
  i64 len, start, j, k, t, zeta;
  k = (size/nsize) - 1;
  i64 *r = (i64*) malloc(sizeof(i64) * size);
  memcpy(r, poly, sizeof(i64)*size );
  poly_mod(r, size, mod);
  for (len = nsize; len <= size/2; len<<=1) {
    for (start = 0; start < size; start = j + len) {
      zeta = zetas[k--];
      for (j = start; j < start + len; j++) {
        t = r[j];
        r[j] = (t + r[j + len]) % mod;
        if (r[j] < 0) r[j] += mod;
        r[j + len] = (r[j + len] - t) % mod;
        if (r[j + len] < 0) r[j + len] += mod;
        r[j + len] = big_mult_mod(zeta, r[j + len], mod);
      }
    }
  }

  i64 invsize = inverse_mod(size/nsize, mod);
  for (j = 0; j < size; j++)
    r[j] = big_mult_mod(r[j], invsize, mod);
  poly_mod(r, size, mod);
  return r;
}

void testmul(i64* r, i64* poly_a, i64* poly_b, i64 mod, i64 zeta, u64 size) {
  i64 i, j;

  for (i = 1; i < size; i++) {
    for (j = size-i; j < size; j++) {
      r[i+j-size] += big_mult_mod(poly_a[i], poly_b[j], mod);
    }
  }

  for (i = 0; i < size; i++) {
    r[i] = big_mult_mod(r[i], zeta, mod);
  }

  for (i = 0; i < size; i++) {
    for (j = 0; j < size-i; j++) {
      r[i+j] += big_mult_mod(poly_a[i], poly_b[j], mod);
    }
  }

  poly_mod(r, size, mod);
}

i64* zmult(i64* poly_a, i64* poly_b, u64 size, i64 mod, i64* metas, u64 nsize) {
  i64 *prod = (i64*) malloc(sizeof(i64) * size);
  memset(prod, 0, sizeof(i64)*size);
  i64 i;
  for (i = 0; i < size/nsize; i++) {
    // basemul(&(prod[nsize*i]), &(poly_a[nsize*i]), &(poly_b[nsize*i]), mod, metas[i]);
    testmul(&(prod[nsize*i]), &(poly_a[nsize*i]), &(poly_b[nsize*i]), mod, metas[i], nsize);
  }

  return prod;
}